File: qt/gen_qshortcut.go
package qt

import (
	"unsafe"
)

type QShortcut struct {
	h          uintptr
	isSubclass bool
}

// NewQShortcut constructs a new QShortcut object.
func NewQShortcut(parent *QObject) *QShortcut {
	ret := newQShortcut(QShortcut_new(parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQShortcut2 constructs a new QShortcut object.
func NewQShortcut2(key *QKeySequence, parent *QObject) *QShortcut {
	ret := newQShortcut(QShortcut_new2(key.cPointer(), parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQShortcut3 constructs a new QShortcut object.
func NewQShortcut3(key QKeySequence__StandardKey, parent *QObject) *QShortcut {
	ret := newQShortcut(QShortcut_new3((int)(key), parent.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQShortcut4 constructs a new QShortcut object.
func NewQShortcut4(key *QKeySequence, parent *QObject, member string) *QShortcut {
	member_Cstring := CString(member)
	defer free(unsafe.Pointer(member_Cstring))

	ret := newQShortcut(QShortcut_new4(key.cPointer(), parent.cPointer(), member_Cstring))
	ret.isSubclass = true
	return ret
}

// NewQShortcut5 constructs a new QShortcut object.
func NewQShortcut5(key *QKeySequence, parent *QObject, member string, ambiguousMember string) *QShortcut {
	member_Cstring := CString(member)
	defer free(unsafe.Pointer(member_Cstring))
	ambiguousMember_Cstring := CString(ambiguousMember)
	defer free(unsafe.Pointer(ambiguousMember_Cstring))

	ret := newQShortcut(QShortcut_new5(key.cPointer(), parent.cPointer(), member_Cstring, ambiguousMember_Cstring))
	ret.isSubclass = true
	return ret
}

// NewQShortcut6 constructs a new QShortcut object.
func NewQShortcut6(key *QKeySequence, parent *QObject, member string, ambiguousMember string, context ShortcutContext) *QShortcut {
	member_Cstring := CString(member)
	defer free(unsafe.Pointer(member_Cstring))
	ambiguousMember_Cstring := CString(ambiguousMember)
	defer free(unsafe.Pointer(ambiguousMember_Cstring))

	ret := newQShortcut(QShortcut_new6(key.cPointer(), parent.cPointer(), member_Cstring, ambiguousMember_Cstring, (int)(context)))
	ret.isSubclass = true
	return ret
}

// NewQShortcut7 constructs a new QShortcut object.
func NewQShortcut7(key QKeySequence__StandardKey, parent *QObject, member string) *QShortcut {
	member_Cstring := CString(member)
	defer free(unsafe.Pointer(member_Cstring))

	ret := newQShortcut(QShortcut_new7((int)(key), parent.cPointer(), member_Cstring))
	ret.isSubclass = true
	return ret
}

// NewQShortcut8 constructs a new QShortcut object.
func NewQShortcut8(key QKeySequence__StandardKey, parent *QObject, member string, ambiguousMember string) *QShortcut {
	member_Cstring := CString(member)
	defer free(unsafe.Pointer(member_Cstring))
	ambiguousMember_Cstring := CString(ambiguousMember)
	defer free(unsafe.Pointer(ambiguousMember_Cstring))

	ret := newQShortcut(QShortcut_new8((int)(key), parent.cPointer(), member_Cstring, ambiguousMember_Cstring))
	ret.isSubclass = true
	return ret
}

// NewQShortcut9 constructs a new QShortcut object.
func NewQShortcut9(key QKeySequence__StandardKey, parent *QObject, member string, ambiguousMember string, context ShortcutContext) *QShortcut {
	member_Cstring := CString(member)
	defer free(unsafe.Pointer(member_Cstring))
	ambiguousMember_Cstring := CString(ambiguousMember)
	defer free(unsafe.Pointer(ambiguousMember_Cstring))

	ret := newQShortcut(QShortcut_new9((int)(key), parent.cPointer(), member_Cstring, ambiguousMember_Cstring, (int)(context)))
	ret.isSubclass = true
	return ret
}

func (this *QShortcut) MetaObject() *QMetaObject {
	return newQMetaObject(QShortcut_MetaObject(this.h))
}

func (this *QShortcut) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QShortcut_Metacast(this.h, param1_Cstring))
}

func QShortcut_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QShortcut_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QShortcut) SetKey(key *QKeySequence) {
	QShortcut_SetKey(this.h, key.cPointer())
}

func (this *QShortcut) Key() *QKeySequence {
	_goptr := newQKeySequence(QShortcut_Key(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QShortcut) SetKeys(key QKeySequence__StandardKey) {
	QShortcut_SetKeys(this.h, (int)(key))
}

func (this *QShortcut) SetKeysWithKeys(keys []QKeySequence) {
	keys_CArray := (*[0xffff]*QKeySequence)(malloc(size_t(8 * len(keys))))
	defer free(unsafe.Pointer(keys_CArray))
	for i := range keys {
		keys_CArray[i] = keys[i].cPointer()
	}
	keys_ma := struct_miqt_array{len: size_t(len(keys)), data: unsafe.Pointer(keys_CArray)}
	QShortcut_SetKeysWithKeys(this.h, keys_ma)
}

func (this *QShortcut) Keys() []QKeySequence {
	var _ma struct_miqt_array = QShortcut_Keys(this.h)
	_ret := make([]QKeySequence, int(_ma.len))
	_outCast := (*[0xffff]*QKeySequence)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		_lv_goptr := newQKeySequence(_outCast[i])
		_lv_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
		_ret[i] = *_lv_goptr
	}
	return _ret
}

func (this *QShortcut) SetEnabled(enable bool) {
	QShortcut_SetEnabled(this.h, (bool)(enable))
}

func (this *QShortcut) IsEnabled() bool {
	return (bool)(QShortcut_IsEnabled(this.h))
}

func (this *QShortcut) SetContext(context ShortcutContext) {
	QShortcut_SetContext(this.h, (int)(context))
}

func (this *QShortcut) Context() ShortcutContext {
	return (ShortcutContext)(QShortcut_Context(this.h))
}

func (this *QShortcut) SetAutoRepeat(on bool) {
	QShortcut_SetAutoRepeat(this.h, (bool)(on))
}

func (this *QShortcut) AutoRepeat() bool {
	return (bool)(QShortcut_AutoRepeat(this.h))
}

func (this *QShortcut) Id() int {
	return (int)(QShortcut_Id(this.h))
}

func (this *QShortcut) SetWhatsThis(text string) {
	text_ms := struct_miqt_string{}
	text_ms.data = CString(text)
	text_ms.len = size_t(len(text))
	defer free(unsafe.Pointer(text_ms.data))
	QShortcut_SetWhatsThis(this.h, text_ms)
}

func (this *QShortcut) WhatsThis() string {
	var _ms struct_miqt_string = QShortcut_WhatsThis(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QShortcut) Activated() {
	QShortcut_Activated(this.h)
}

func (this *QShortcut) OnActivated(slot func()) {
	QShortcut_connect_Activated(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QShortcut_Activated
func miqt_exec_callback_QShortcut_Activated(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QShortcut) ActivatedAmbiguously() {
	QShortcut_ActivatedAmbiguously(this.h)
}

func (this *QShortcut) OnActivatedAmbiguously(slot func()) {
	QShortcut_connect_ActivatedAmbiguously(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QShortcut_ActivatedAmbiguously
func miqt_exec_callback_QShortcut_ActivatedAmbiguously(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QShortcut_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QShortcut_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QShortcut_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QShortcut_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QShortcut) callVirtualBase_MetaObject() *QMetaObject {
	return newQMetaObject(QShortcut_virtualbase_MetaObject(unsafe.Pointer(this.h)))
}

func (this *QShortcut) OnMetaObject(slot func(super func() *QMetaObject) *QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QShortcut_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QShortcut_MetaObject
func miqt_exec_callback_QShortcut_MetaObject(self QShortcut, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *QMetaObject) *QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QShortcut{h: self}).callVirtualBase_MetaObject)

	return virtualReturn.cPointer()
}

func (this *QShortcut) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QShortcut_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QShortcut) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QShortcut_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QShortcut_Metacast
func miqt_exec_callback_QShortcut_Metacast(self QShortcut, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QShortcut{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/gen_qshortcut.cpp
// +build ignore

#include <QEvent>
#include <QKeySequence>
#include <QList>
#include <QMetaObject>
#include <QObject>
#include <QShortcut>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qshortcut.h>
#include "gen_qshortcut.h"
class MiqtVirtualQShortcut : public virtual QShortcut {
public:
MiqtVirtualQShortcut(QObject* parent): QShortcut(parent) {};
MiqtVirtualQShortcut(const QKeySequence& key, QObject* parent): QShortcut(key, parent) {};
MiqtVirtualQShortcut(QKeySequence::StandardKey key, QObject* parent): QShortcut(key, parent) {};
MiqtVirtualQShortcut(const QKeySequence& key, QObject* parent, const char* member): QShortcut(key, parent, member) {};
MiqtVirtualQShortcut(const QKeySequence& key, QObject* parent, const char* member, const char* ambiguousMember): QShortcut(key, parent, member, ambiguousMember) {};
MiqtVirtualQShortcut(const QKeySequence& key, QObject* parent, const char* member, const char* ambiguousMember, Qt::ShortcutContext context): QShortcut(key, parent, member, ambiguousMember, context) {};
MiqtVirtualQShortcut(QKeySequence::StandardKey key, QObject* parent, const char* member): QShortcut(key, parent, member) {};
MiqtVirtualQShortcut(QKeySequence::StandardKey key, QObject* parent, const char* member, const char* ambiguousMember): QShortcut(key, parent, member, ambiguousMember) {};
MiqtVirtualQShortcut(QKeySequence::StandardKey key, QObject* parent, const char* member, const char* ambiguousMember, Qt::ShortcutContext context): QShortcut(key, parent, member, ambiguousMember, context) {};

virtual ~MiqtVirtualQShortcut() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QShortcut::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QShortcut_MetaObject(const_cast<MiqtVirtualQShortcut*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QShortcut::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QShortcut::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QShortcut_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QShortcut::qt_metacast(param1);

	}
};
QShortcut* QShortcut_new(QObject* parent) {
return new MiqtVirtualQShortcut(parent);
}
QShortcut* QShortcut_new2(QKeySequence* key, QObject* parent) {
return new MiqtVirtualQShortcut(*key, parent);
}
QShortcut* QShortcut_new3(int key, QObject* parent) {
return new MiqtVirtualQShortcut(static_cast<QKeySequence::StandardKey>(key), parent);
}
QShortcut* QShortcut_new4(QKeySequence* key, QObject* parent, const char* member) {
return new MiqtVirtualQShortcut(*key, parent, member);
}
QShortcut* QShortcut_new5(QKeySequence* key, QObject* parent, const char* member, const char* ambiguousMember) {
return new MiqtVirtualQShortcut(*key, parent, member, ambiguousMember);
}
QShortcut* QShortcut_new6(QKeySequence* key, QObject* parent, const char* member, const char* ambiguousMember, int context) {
return new MiqtVirtualQShortcut(*key, parent, member, ambiguousMember, static_cast<Qt::ShortcutContext>(context));
}
QShortcut* QShortcut_new7(int key, QObject* parent, const char* member) {
return new MiqtVirtualQShortcut(static_cast<QKeySequence::StandardKey>(key), parent, member);
}
QShortcut* QShortcut_new8(int key, QObject* parent, const char* member, const char* ambiguousMember) {
return new MiqtVirtualQShortcut(static_cast<QKeySequence::StandardKey>(key), parent, member, ambiguousMember);
}
QShortcut* QShortcut_new9(int key, QObject* parent, const char* member, const char* ambiguousMember, int context) {
return new MiqtVirtualQShortcut(static_cast<QKeySequence::StandardKey>(key), parent, member, ambiguousMember, static_cast<Qt::ShortcutContext>(context));
}
void QShortcut_virtbase(QShortcut* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QShortcut_MetaObject(const QShortcut* self) {
	return (QMetaObject*) self->metaObject();
}
void* QShortcut_Metacast(QShortcut* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QShortcut_Tr(const char* s) {
	QString _ret = QShortcut::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QShortcut_SetKey(QShortcut* self, QKeySequence* key) {
	self->setKey(*key);
}
QKeySequence* QShortcut_Key(const QShortcut* self) {
	return new QKeySequence(self->key());
}
void QShortcut_SetKeys(QShortcut* self, int key) {
	self->setKeys(static_cast<QKeySequence::StandardKey>(key));
}
void QShortcut_SetKeysWithKeys(QShortcut* self, struct miqt_array /* of QKeySequence* */  keys) {
	QList<QKeySequence> keys_QList;
	keys_QList.reserve(keys.len);
	QKeySequence** keys_arr = static_cast<QKeySequence**>(keys.data);
	for(size_t i = 0; i < keys.len; ++i) {
		keys_QList.push_back(*(keys_arr[i]));
	}
	self->setKeys(keys_QList);
}
struct miqt_array /* of QKeySequence* */  QShortcut_Keys(const QShortcut* self) {
	QList<QKeySequence> _ret = self->keys();
	// Convert QList<> from C++ memory to manually-managed C memory
	QKeySequence** _arr = static_cast<QKeySequence**>(malloc(sizeof(QKeySequence*) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		_arr[i] = new QKeySequence(_ret[i]);
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}
void QShortcut_SetEnabled(QShortcut* self, bool enable) {
	self->setEnabled(enable);
}
bool QShortcut_IsEnabled(const QShortcut* self) {
	return self->isEnabled();
}
void QShortcut_SetContext(QShortcut* self, int context) {
	self->setContext(static_cast<Qt::ShortcutContext>(context));
}
int QShortcut_Context(const QShortcut* self) {
	Qt::ShortcutContext _ret = self->context();
	return static_cast<int>(_ret);
}
void QShortcut_SetAutoRepeat(QShortcut* self, bool on) {
	self->setAutoRepeat(on);
}
bool QShortcut_AutoRepeat(const QShortcut* self) {
	return self->autoRepeat();
}
int QShortcut_Id(const QShortcut* self) {
	return self->id();
}
void QShortcut_SetWhatsThis(QShortcut* self, struct miqt_string text) {
	QString text_QString = QString::fromUtf8(text.data, text.len);
	self->setWhatsThis(text_QString);
}
struct miqt_string QShortcut_WhatsThis(const QShortcut* self) {
	QString _ret = self->whatsThis();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QShortcut_Activated(QShortcut* self) {
	self->activated();
}
void QShortcut_connect_Activated(QShortcut* self, intptr_t slot) {
	MiqtVirtualQShortcut::connect(self, static_cast<void (QShortcut::*)()>(&QShortcut::activated), self, [=]() {
		miqt_exec_callback_QShortcut_Activated(slot);
	});
}
void QShortcut_ActivatedAmbiguously(QShortcut* self) {
	self->activatedAmbiguously();
}
void QShortcut_connect_ActivatedAmbiguously(QShortcut* self, intptr_t slot) {
	MiqtVirtualQShortcut::connect(self, static_cast<void (QShortcut::*)()>(&QShortcut::activatedAmbiguously), self, [=]() {
		miqt_exec_callback_QShortcut_ActivatedAmbiguously(slot);
	});
}
struct miqt_string QShortcut_Tr2(const char* s, const char* c) {
	QString _ret = QShortcut::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QShortcut_Tr3(const char* s, const char* c, int n) {
	QString _ret = QShortcut::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QShortcut_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQShortcut*>( (QShortcut*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QShortcut_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQShortcut*)(self) )->virtualbase_MetaObject();
}
void QShortcut_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQShortcut*>( (QShortcut*)(self) )->handle__Metacast = slot;
}
void* QShortcut_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQShortcut*)(self) )->virtualbase_Metacast(param1);
}
void QShortcut_Delete(QShortcut* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQShortcut*>( self );
	} else {
		delete self;
	}
}
